docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,247 @@
+
+Global to Zone synchronize resources:
+* CircuitBreaker
+* ExternalService
+* FaultInjection
+* HealthCheck
+* Mesh
+* MeshGateway
+* MeshGatewayRoute
+* ProxyTemplate
+* RateLimit
+* Retry
+* ServiceInsight
+* Timeout
+* TrafficLog
+* TrafficPermission
+* TrafficRoute
+* TrafficTrace
+* VirtualOutbound
+* ZoneIngress*
+* Secret
+* Config
+* MeshAccessLog
+* MeshCircuitBreaker
+* MeshFaultInjection
+* MeshHealthCheck
+* MeshHttpRoute
+* MeshProxyPatch
+* MeshRateLimit
+* MeshRetry
+* MeshTimeout
+* MeshTrace
+* MeshTrafficPermission
+
+A client sends DiscoveryRequest to the server and waits for the change. The server stores the state in a state-of-the-world cache and sends the response whenever there is a change. It works well in the case of Global to Zone synchronization
+but might have scalability issues for synchronization Zone to Global with large deployments. Zone control-plane needs to send all the Dataplane resources to the global whenever there is a change. That might be a bottleneck that can happen in large deployments. We can use [`Incremental xDS`](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol) which is Envoy's implementation of cache that instead of sending all the resources sends only changes. That could reduce the bandwidth usage.
+
+```proto
+service GlobalKDSService {
+ ...
+ rpc ZoneToGlobalSync(stream envoy.service.discovery.v3.DeltaDiscoveryRequest)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryResponse);
+
+ rpc GlobalToZoneSync(stream envoy.service.discovery.v3.DeltaDiscoveryRequest)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryRequest);
+}
+
+```
+
+In the first iteration, we can introduce only `ZoneToGlobalSync` with delta xDS, and in the future, we could introduce `GlobalToZoneSync` and also use delta xDS. In the code, I've noticed that we are using our copy of the Snapshot cache and the code is a bit outdated. As a part of that task, we should check if we can change to use `go-control-plane` cache implementation or copy and update our current implementation. 
+
+Maybe not in the first step but we could consider if we can benefit from using different cache implementation `LinearCache` or `MuxCache` for snapshot generation.
+
+### Prioritize operational informations
+
+Certain resources require more frequent synchronization than others. For example, Policies, ZoneIngress, and ZoneEgress must be updated immediately, whereas Insights and Dataplanes objects may be delayed and can be aggregated. To address this, we could consider implementing an option to specify the interval at which the snapshot for specific resources is recalculated. To avoid recalculating changes for all synchronized resources, we need a cache per resource type, which `SimpleCache` should allows us to achieve. In addition, we could set up an RPC service to monitor changes for all resources on separate streams, allowing changes to be processed independently.
```

Comment:
Doesn't the cache get invalidated when resources change?

Reply:
Yes, but delta xDS keeps a map of typeURL: resourceName - version, so based on this it can make to correct calculation and sends only the changed resource. https://github.com/envoyproxy/go-control-plane/blob/a51ba6fa54b2e107f5606a0e2799f8afed116c14/pkg/cache/v3/snapshot.go#L35